src/frame_buffer/single_frame_allocator.rs

```diff
@@ -63,49 +51,44 @@ impl SingleFrameAllocator {
     ///   let allocator = SingleFrameAllocator::new();
     /// # }
     /// ```
-    pub fn new() -> Self {
-        let frame_allocator = Self {
-            is_allocated: false,
-        };
-
-        crit_sect::locked(|cs| {
-            let prev_frame_allocator = FRAME_ALLOCATOR.borrow(cs).replace(Some(frame_allocator));
-            assert!(prev_frame_allocator.is_none());
-        });
-
-        let phantom_allocator = Self { is_allocated: true };
-        phantom_allocator
+    pub const fn new() -> Self {
+        Self {
+            is_allocated: AtomicBool::new(false),
+            data: UnsafeCell::new([0; Self::BUF_SIZE]),
+        }
     }
 
     /// Releases an allocated frame
     ///
     /// Function called when any [`FrameBuffer`](super::frame_buffer::FrameBuffer) created by this
     /// allocator is dropped.
-    fn release_frame(_frame: &mut FrameBuffer, _metadata: DropMetadata) {
-        SingleFrameAllocator::use_frame_allocator(|fa| {
-            assert!(fa.is_allocated);
-            fa.is_allocated = false;
-        });
+    pub fn release_frame(&self, _frame: &mut FrameBuffer) {
+        self.is_allocated
+            .store(false, core::sync::atomic::Ordering::Relaxed);
     }
 }
 
+unsafe impl Sync for SingleFrameAllocator {}
+
 impl FrameAllocator for SingleFrameAllocator {
     fn get_frame(&self) -> Result<FrameBuffer, Error> {
-        let mut result = Err(Error::WouldBlock);
-        SingleFrameAllocator::use_frame_allocator(|fa| {
-            result = if fa.is_allocated {
-                Err(Error::WouldBlock)
-            } else {
-                fa.is_allocated = true;
-                Ok(FrameBuffer::new(
-                    unsafe { &mut DATA },
-                    Some(SingleFrameAllocator::release_frame),
-                    &None::<usize>,
-                ))
-            }
-        });
-
-        result
+        if self
+            .is_allocated
+            .compare_exchange(
+                false,
+                true,
+                core::sync::atomic::Ordering::Release,
+                core::sync::atomic::Ordering::Relaxed,
+            )
+            .is_ok()
+        {
+            Ok(FrameBuffer::new(
+                FrameAllocators::SingleFrameAllocator(self),
+                unsafe { &mut *self.data.get() },
```

Comment:
A short comment why `unsafe` code is sound here would be nice. I think I get it, but explicit statement wouldn't harm :)

Reply:
Yes, I forgot about it...